### ISSUE TITLE
Fix S4158 FP: rule raised where there is no guarantees of emptiness

### ISSUE THREAD

**Issue by rjgotten** [SonarSource/sonar-dotnet#7582]
### Description

S4158 regarding iteration of known empty collections may be raised in situations where there is no guarantee that the collection is empty.

### Repro steps

```cs
private static void Demo(IEnumerable<int> input)
{
  var list = new List<int>();
  var flag = false;

  foreach(var item in input)
  {
    if (item > 10)
    {
      flag = true;
      continue;
    }

    list.Add(item);
  }

  if (flag)
  {
    foreach(var item in list)
    {
      // do something with items that were <= 10
      // only when items > 10 were found.
    }
  }
}
```

### Expected behavior

S4158 is not raised, because there is no guarantee that `list` will be empty when `flag` is true.

### Actual behavior

S4158 is raised.

### Known workarounds

Suppress the rule.

### Related information

* C#/VB.NET Plugins version : C# 9.4.0.72892 (SonarLint  7.0.0.74072)
* Visual Studio version: 17.6.4
* MSBuild / dotnet version: dotnet 7.0.304
* Operating System: Windows 11

**Comment by Tim-Pohlmann** [comment]
Thanks @rjgotten, for reporting this. I confirm it as an FP.
